crates/core/src/buc/controller.rs

```diff
@@ -0,0 +1,334 @@
+use crate::{
+	ctx::{Context, MutableContext},
+	dbs::Options,
+	doc::CursorDoc,
+	err::Error,
+	iam::Action,
+	sql::{statements::define::BucketDefinition, Bytes, File, FlowResultExt, Permission, Value},
+};
+use core::fmt;
+use reblessive::tree::Stk;
+use std::sync::Arc;
+
+use super::store::{Key, ListOptions, ObjectMeta, ObjectStore};
+
+/// Allows you to control a specific bucket in the context of the current user
+pub struct BucketController<'a> {
+	stk: &'a mut Stk,
+	ctx: &'a Context,
+	opt: &'a Options,
+	doc: Option<&'a CursorDoc>,
+
+	bucket: Arc<BucketDefinition>,
+	store: Arc<dyn ObjectStore>,
+}
+
+impl<'a> BucketController<'a> {
+	/// Create a `FileController` for a specified file
+	/// Will obtain a bucket connection and return back a `FileController` or `Error`
+	pub(crate) async fn new(
+		stk: &'a mut Stk,
+		ctx: &'a Context,
+		opt: &'a Options,
+		doc: Option<&'a CursorDoc>,
+		buc: &str,
+	) -> Result<Self, Error> {
+		let (ns, db) = opt.ns_db()?;
+		let bucket = ctx.tx().get_db_bucket(ns, db, buc).await?;
+		let store = ctx.get_bucket_store(ns, db, buc).await?;
+
+		Ok(Self {
+			stk,
+			ctx,
+			opt,
+			doc,
+
+			bucket,
+			store,
+		})
+	}
+
+	/// Checks if the bucket allows writes, and if not, return an `Error::ReadonlyBucket`
+	fn require_writeable(&self) -> Result<(), Error> {
+		if self.bucket.readonly {
+			Err(Error::ReadonlyBucket(self.bucket.name.to_raw()))
+		} else {
+			Ok(())
+		}
+	}
+
+	/// Attempt to put a file
+	/// `Bytes` and `Strand` values are supported, and will be converted into `Bytes`
+	/// Create or update permissions will be used, based on if the remote file already exists
+	pub(crate) async fn put(&mut self, key: &Key, value: Value) -> Result<(), Error> {
+		let payload = match value {
+			Value::Bytes(v) => v.0.into(),
+			Value::Strand(v) => v.0.into_bytes().into(),
+			from => {
+				return Err(Error::ConvertTo {
+					from,
+					into: "bytes".into(),
+				})
+			}
+		};
+
+		self.require_writeable()?;
+		self.check_permission(BucketOperation::Put, Some(key), None).await?;
+
+		self.store
+			.put(key, payload)
+			.await
+			.map_err(|e| Error::ObjectStoreFailure(self.bucket.name.to_raw(), e.to_string()))?;
+
+		Ok(())
+	}
+
+	/// Attempt to put a file
+	/// `Bytes` and `Strand` values are supported, and will be converted into `Bytes`
+	/// Create or update permissions will be used, based on if the remote file already exists
+	pub(crate) async fn put_if_not_exists(&mut self, key: &Key, value: Value) -> Result<(), Error> {
+		let payload = match value {
+			Value::Bytes(v) => v.0.into(),
+			Value::Strand(v) => v.0.into_bytes().into(),
+			from => {
+				return Err(Error::ConvertTo {
+					from,
+					into: "bytes".into(),
+				})
+			}
+		};
```

Comment:
This code is duplicated with the previous function. Can we factorize the code?

Reply:
Done in [120e55e](https://github.com/surrealdb/surrealdb/pull/5701/commits/120e55e0ad47bb91671a7d96931d22c5488e2b96)